### DISCUSSION TITLE
The problem of user data

### DISCUSSION THREAD

**Discussion by Silence-zyx** [microsoft/CCF#2838] [Q&A] [answered]
There is a step to set user data at [Opening a network](https://microsoft.github.io/CCF/ccf-1.0.6/governance/open_network.html), and `user_id` is required.
I have no idea how to get it. So, I just use a random (64 bit)  as the parameter to generate the proposal with python.
`$ cat set_user_data_proposal.json`
`{`
`    "actions": [`
`        {`
`            "name": "set_user_data",`
`            "args": {`
`                "user_id": "1517904000521082707645555088587415055510748411558110628298834154",`
`                "user_data": {`
`                    "isAdmin": true`
`                }`
`           }`
`        }`
`    ]`
`}`

And I add the user to the network and open it.
The command
`curl https://127.0.0.1:8000/app/commit \`
`-X GET \`
`--key  ....../CCF/ccf-1.0.6/CCF-ccf-1.0.6/python/utils/user0_privk.pem \`
`--cert ......./CCF/ccf-1.0.6/CCF-ccf-1.0.6/python/utils/user0_cert.pem \`
`--cacert /opt/ccf/bin/networkcert.pem`
returns
`{"transaction_id":"2.18"}`
It seems good~
And I set the js scripts with commands
`python3.8 ./proposal_generator.py \`
`--pretty-print \`
`--proposal-output-file set_js_app_proposal.json \`
`--vote-output-file set_js_app_vote_for.json \`
`set_js_app ......./CCF/ccf-1.0.6/CCF-ccf-1.0.6/samples/apps/logging/js`
and vote it.

However, when I type this command,
`curl https://127.0.0.1:8000/app/log/private \`
`-X POST \`
`--key /home/yongxin/CCF/ccf-1.0.6/CCF-ccf-1.0.6/python/utils/user0_privk.pem \`
`--cert /home/yongxin/CCF/ccf-1.0.6/CCF-ccf-1.0.6/python/utils/user0_cert.pem \`
`--cacert /opt/ccf/bin/networkcert.pem \`
`-H "Content-Type: application/json" \`
`--data-binary '{"id": 42, "msg": "Logged to private table"}' \`
`-i`
it returns
`HTTP/1.1 401 Unauthorized`
`content-length: 99`
`content-type: application/json`
`{"error":{"code":"InvalidAuthenticationInfo","message":"Could not find matching user certificate"}}` 

So, what caused this result?
The id of the user0?  The JS script is not right?  Or the user is not in the network?

By the way, is there a command to obtain the current user list of the CCF network?

**Comment by achamayou** [comment, marked as answer]
@Silence-zyx please see https://microsoft.github.io/CCF/main/governance/open_network.html#adding-users for how to add users to a network.

Setting user data is only attaching additional user data to a user id, it is not creating the user identity itself.

**Reply by Silence-zyx** [reply]
Thank you, sir. 
But I have already added the user to the network, that is why I can get the `{"transaction_id": "2.18"}` when I type `curl https://127.0.0.1:8000/app/commit ......` with user0's `.pem` file. However, I still get the problem`"message":"Could not find matching user certificate"`  mentioned above.

**Reply by Silence-zyx** [reply]
Of course, the proposal state of adding user is accepted

**Reply by achamayou** [reply]
`/app/commit` is unauthenticated: https://github.com/microsoft/CCF/blob/ad8d65091e8bd75d374aafca976fef47c3b2ec12/src/endpoints/common_endpoint_registry.cpp#L88
so the client cert supplied is not checked. `/app/log/private` on the other hand requires either jwt or client cert auth: https://github.com/microsoft/CCF/blob/ad8d65091e8bd75d374aafca976fef47c3b2ec12/samples/apps/logging/logging.cpp#L186

Can you confirm that you did add a user with this cert? Note that user ids are allocated by CCF, and are the sha-256 digest of the supplied certificate, so it's unlikely that calling set_user_data with an arbitrary user id value is doing what you expect.

**Reply by Silence-zyx** [reply]
I reset the user id. I use `$ sha256sum user0_cert.pem` to calculate the id.  But it still the problem `"message":"Could not find matching user certificate"`
